<a href="https://colab.research.google.com/github/trevorlillywhite/HW-4---NCSU-ST-554---Trevor-Lillywhite/blob/main/Task_2_Project_1_(Lead_Trevor).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Analysis of Big Data - NCSU ST 554
## Project 1
### Task 2 - Exploratory Data Analysis

### Group Members:
+ Trevor Lillywhite (primary contributor to Task 2)
+ Cass Crews
+ Joy Zhou

#### Due Date: 2/23 (soft), 2/25 (hard)

##### Task Overview:

The overall objective of this notebook is to conduct extensive Exploratory Data Analysis (EDA) on a dataset to understand relationships between sensor and weather variables vs. the concentration of benzene.

The dataset of interest is the "Air Quality" dataset from the UC Irvine Machine Learning Repository, available here:  
+ https://archive.ics.uci.edu/dataset/360/air+quality
This data consists of time series data (data recorded over time) for air quality measurements in Italy, as described in this paper:
+ https://www.semanticscholar.org/paper/On-field-calibration-of-an-electronic-nose-for-in-Vito-Massera/a90a54a39ff934772df57771a0012981f355949d

This notebook will include many EDA elements, including (but not limited to):
+ Numeric summaries of the benzene (`C6H6 (GT)` variable at different levels/combinations of other variables
+ Correlations between variables
+ Plots of the `C6H6 (GT)` variable, showing relationships with other variables
+ Relationships over time or ignoring time.

Sufficient narrative will be provided to give the reader insights into the thought processes and conclusions reached from EDA.

#### Read in Dataset

First, we will import relevant `python` libraries to enable our subsequent analysis.

In [2]:
# Import standard data analysis libraries
import pandas as pd                     # Convenient data structuring (DataFrames)
import numpy as np                      # Numerical tools (e.g., arrays)
from matplotlib import pyplot as plt    # Plotting tools
import seaborn as sns                   # Alternative plotting tools

Next, we will install the UCI Machine Learning Repository via a `pip install` command and then fetch the specific dataset of interest.

In [3]:
!pip install ucimlrepo      # Installs repository of datasets
import ucimlrepo as uci     # Imports newly installed library
air_quality = uci.fetch_ucirepo(id=360)     # Fetches dataset of interest

Now that we have a variable `air_quality` with the desired dataset, let's look at the data! We will start with looking at object type and then using the `features` attribute to tell us the column names. Then we will look at some basic statistics and example rows (observations) before making any changes.  

In [4]:
# Show object data type
type(air_quality)

ucimlrepo.dotdict.dotdict

In [5]:
# Show column names (features)
air_quality.data.features

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0     3/10/2004  18:00:00     2.6         1360       150      11.9   
1     3/10/2004  19:00:00     2.0         1292       112       9.4   
2     3/10/2004  20:00:00     2.2         1402        88       9.0   
3     3/10/2004  21:00:00     2.2         1376        80       9.2   
4     3/10/2004  22:00:00     1.6         1272        51       6.5   
...         ...       ...     ...          ...       ...       ...   
9352   4/4/2005  10:00:00     3.1         1314      -200      13.5   
9353   4/4/2005  11:00:00     2.4         1163      -200      11.4   
9354   4/4/2005  12:00:00     2.4         1142      -200      12.4   
9355   4/4/2005  13:00:00     2.1         1003      -200       9.5   
9356   4/4/2005  14:00:00     2.2         1071      -200      11.9   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
0              1046      166          1056      113          1692   
1               955      103          1174       92          1559   
2               939      131          1140      114          1555   
3               948      172          1092      122          1584   
4               836      131          1205      116          1490   
...             ...      ...           ...      ...           ...   
9352           1101      472           539      190          1374   
9353           1027      353           604      179          1264   
9354           1063      293           603      175          1241   
9355            961      235           702      156          1041   
9356           1047      265           654      168          1129   

      PT08.S5(O3)     T    RH      AH  
0            1268  13.6  48.9  0.7578  
1             972  13.3  47.7  0.7255  
2            1074  11.9  54.0  0.7502  
3            1203  11.0  60.0  0.7867  
4            1110  11.2  59.6  0.7888  
...           ...   ...   ...     ...  
9352         1729  21.9  29.3  0.7568  
9353         1269  24.3  23.7  0.7119  
9354         1092  26.9  18.3  0.6406  
9355          770  28.3  13.5  0.5139  
9356          816  28.5  13.1  0.5028  

[9357 rows x 15 columns]

This appears to be a dictionary-type object with 15 columns and 9357 rows.

To use some convenient data exploration features, we will convert this into a Pandas DataFrame.

In [10]:
df_air_quality = pd.DataFrame(air_quality.data.features)    # Convert to DataFrame
df_air_quality.info()                                       # View basic info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   int64  
 4   NMHC(GT)       9357 non-null   int64  
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   int64  
 7   NOx(GT)        9357 non-null   int64  
 8   PT08.S3(NOx)   9357 non-null   int64  
 9   NO2(GT)        9357 non-null   int64  
 10  PT08.S4(NO2)   9357 non-null   int64  
 11  PT08.S5(O3)    9357 non-null   int64  
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
dtypes: float64(5), int64(8), object(2)
memory usage: 1.1+ MB


All columns and rows were converted to a DataFrame data structure. Data types include `float64`, `int64`, and `object`.

Next, we will use the `describe()` method to show basic statistics on numeric-type columns.

In [11]:
df_air_quality.describe()

CO(GT)  PT08.S1(CO)     NMHC(GT)     C6H6(GT)  PT08.S2(NMHC)  \
count  9357.000000  9357.000000  9357.000000  9357.000000    9357.000000   
mean    -34.207524  1048.990061  -159.090093     1.865683     894.595276   
std      77.657170   329.832710   139.789093    41.380206     342.333252   
min    -200.000000  -200.000000  -200.000000  -200.000000    -200.000000   
25%       0.600000   921.000000  -200.000000     4.000000     711.000000   
50%       1.500000  1053.000000  -200.000000     7.900000     895.000000   
75%       2.600000  1221.000000  -200.000000    13.600000    1105.000000   
max      11.900000  2040.000000  1189.000000    63.700000    2214.000000   

           NOx(GT)  PT08.S3(NOx)      NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
count  9357.000000   9357.000000  9357.000000   9357.000000  9357.000000   
mean    168.616971    794.990168    58.148873   1391.479641   975.072032   
std     257.433866    321.993552   126.940455    467.210125   456.938184   
min    -200.000000   -200.000000  -200.000000   -200.000000  -200.000000   
25%      50.000000    637.000000    53.000000   1185.000000   700.000000   
50%     141.000000    794.000000    96.000000   1446.000000   942.000000   
75%     284.000000    960.000000   133.000000   1662.000000  1255.000000   
max    1479.000000   2683.000000   340.000000   2775.000000  2523.000000   

                 T           RH           AH  
count  9357.000000  9357.000000  9357.000000  
mean      9.778305    39.485380    -6.837604  
std      43.203623    51.216145    38.976670  
min    -200.000000  -200.000000  -200.000000  
25%      10.900000    34.100000     0.692300  
50%      17.200000    48.600000     0.976800  
75%      24.100000    61.900000     1.296200  
max      44.600000    88.700000     2.231000

Our final step before cleaning the data will be to show a few example rows using the `head()` method to see precisely what the data looks like, especially the `object` type columns (Date and Time).

In [12]:
df_air_quality.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  3/10/2004  18:00:00     2.6         1360       150      11.9   
1  3/10/2004  19:00:00     2.0         1292       112       9.4   
2  3/10/2004  20:00:00     2.2         1402        88       9.0   
3  3/10/2004  21:00:00     2.2         1376        80       9.2   
4  3/10/2004  22:00:00     1.6         1272        51       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0           1046      166          1056      113          1692         1268   
1            955      103          1174       92          1559          972   
2            939      131          1140      114          1555         1074   
3            948      172          1092      122          1584         1203   
4            836      131          1205      116          1490         1110   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888